include/clang/StaticAnalyzer/Core/BugReporter/InternalInfo.h

```diff
@@ -0,0 +1,180 @@
+    return Info.getKind() == getStaticKind();
+  }
+
+  /// Profile - Used to profile the contents of this object for inclusion in a
+  /// FoldingSet.
+  void Profile(llvm::FoldingSetNodeID &ID) const {
+    ID.AddInteger(static_cast<int>(kind));
+  }
+
+  template <typename T> T *getAs() {
+    return T::isKind(*this) ? static_cast<T *>(this) : nullptr;
+  }
+
+  using StringPrinter = llvm::raw_ostream &(llvm::raw_ostream &, StringRef);
+
+  virtual void print(llvm::raw_ostream &Out, StringPrinter Transform,
+                     StringRef NewlinePrefix = "",
+                     StringRef NewlinePostfix = "\n") const = 0;
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+  LLVM_DUMP_METHOD void dump() const { dumpToStream(llvm::errs()); }
+  LLVM_DUMP_METHOD virtual void dumpToStream(llvm::raw_ostream &Out) const = 0;
+#endif
+};
+
+class InternalInfoMap {
+  using InternalInfoMapImpl = llvm::FoldingSet<InternalInfo>;
+
+public:
+  using iterator = InternalInfoMapImpl::iterator;
+  using const_iterator = InternalInfoMapImpl::const_iterator;
+  using value_type = InternalInfo;
+
+private:
+  llvm::FoldingSet<InternalInfo> Set;
+
+public:
+  bool empty() const { return Set.empty(); }
+  iterator begin() { return Set.begin(); }
+  const_iterator begin() const { return Set.begin(); }
+  iterator end() { return Set.end(); }
+  const_iterator end() const { return Set.end(); }
+
+  /// Get an InternalInfo object from the map. If the map doesn't contain an
+  /// object of type \p T, it inserts it first.
+  ///
+  /// \param T - Descendant of InternalInfo.
+  template <typename T, typename... Args> T &getOrInsert(Args... CtorArgs) {
+    static_assert(std::is_base_of<InternalInfo, T>::value,
+                  "InternalInfoMap may only hold descendants of InternalInfo!");
+
+    llvm::FoldingSetNodeID ID;
+    ID.AddInteger(static_cast<int>(T::getStaticKind()));
+    void *Pos;
+
+    InternalInfo *Val = Set.FindNodeOrInsertPos(ID, Pos);
+    if (!Val) {
+      Val = new T(CtorArgs...);
+      Set.InsertNode(Val, Pos);
+    }
+
+    return static_cast<T &>(*Val);
+  }
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+  LLVM_DUMP_METHOD void dump() const { dumpToStream(llvm::errs()); }
+  LLVM_DUMP_METHOD void dumpToStream(llvm::raw_ostream &Out) const;
+#endif
+};
+
+//===----------------------------------------------------------------------===//
+// Implementations of InternalInfo.
+//===----------------------------------------------------------------------===//
+
+class ConstraintInfo : public InternalInfo {
+public:
+  using StrStrMap = llvm::StringMap<std::string>;
+
+private:
+  /// For debugging reasons, we'll store the constraints imposed on this piece.
+  /// We can't store ConstraintRangeTy directly, as its pointers would be
+  /// invalidated by the time we'd consume this piece.
+  StrStrMap Constraints;
+  /// We'll also store a map of (variable, symbol) pairs (eg. (i, reg_$0))
+  /// pairs.
```

Comment:
`pairs` is repeated twice and a `(` is not closed.

Didn't I comment this earlier?! Make sure you didn't mess up a rebase!

Reply:
Hah, I did: https://github.com/Szelethus/clang/pull/2#discussion_r196736047